### PR TITLE
Added Synchronous option to scale call

### PR DESCRIPTION
**Pull request by vrakesh**
*Issue #, if available:*
Synchronized scale branch call
*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+package com.amazonaws.ml.mms.wlm;
+
+import java.util.concurrent.CompletableFuture;
+
+public class WorkerStateListener {
+
+    private CompletableFuture<Boolean> future;
+    private int successCount;
+    private int failureCount;
+    private int count;
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java]: Use AtomicInt

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    public WorkerStateListener(CompletableFuture<Boolean> future, int count) {
+        this.future = future;
+        this.count = count;
+    }
+
+    public synchronized void notifyOnLoad(boolean success) {
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java]: no synchronized needed
statusChanged(int state)

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 3]
```diff
@@ -114,6 +117,16 @@ public void run() {
                 if (req.getCommand().equals(WorkerCommands.PREDICT)) {
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: We should be able to use ==
case is better

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 4]
```diff
@@ -114,6 +117,16 @@ public void run() {
...
                 }
+                if (req.getCommand().equals(WorkerCommands.LOAD)) {
+                    //success/failure
+                    if (reply.getCode().equals(String.valueOf(200))) {
+                        listener.notifyOnLoad(true);
+                    } else {
+                        listener.notifyOnLoad(false);
+                    }
+                }
+                // success loaded model
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: remove it

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 5]
```diff
@@ -114,6 +117,16 @@ public void run() {
                 if (req.getCommand().equals(WorkerCommands.PREDICT)) {
                     model.resetFailedInfReqs();
                 }
+                if (req.getCommand().equals(WorkerCommands.LOAD)) {
+                    //success/failure
+                    if (reply.getCode().equals(String.valueOf(200))) {
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: use reply.getCode == 200

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 6]
```diff
@@ -127,6 +140,10 @@ public void run() {
             if (req != null) {
                 aggregator.sendError(
                         req, ErrorCodes.INTERNAL_SERVER_ERROR_BACKEND_WORKER_INSTANTIATION);
+
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: You don't need this

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 7]
```diff
@@ -127,6 +140,10 @@ public void run() {
...
+                if (req.getCommand().equals(WorkerCommands.LOAD)) {
+                    listener.notifyOnLoad(false);
+                }
             }
             lifeCycle.exit();
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: listener.statusChanged

File: frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java [hunk 8]
```diff
@@ -345,22 +346,46 @@ private void handleScaleModel(
...
+                            })
+                    .exceptionally(
+                            (e) -> {
+                                NettyUtils.sendError(
+                                        ctx,
+                                        HttpResponseStatus.INTERNAL_SERVER_ERROR,
+                                        e.getMessage());
+                                return null;
+                            });
+
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java]: remove unnecessary empty line

File: frontend/server/src/main/java/com/amazonaws/ml/mms/metrics/MetricCollector.java [hunk 9]
```diff
@@ -70,12 +72,10 @@ public String collect() throws IOException {
         final Process p = Runtime.getRuntime().exec(args, env, workingDir);
         new Thread(
                         () -> {
-                            try {
-                                logger.error(
-                                        IOUtils.toString(
-                                                p.getErrorStream(), StandardCharsets.UTF_8));
-                            } catch (IOException e) {
-                                logger.error(null, e);
+                            InputStream is = p.getErrorStream();
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/metrics/MetricCollector.java]: String str = IOUtils.toString();
if (!str.isEmpty()) {
   logger.error(str);
}

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+    public WorkerStateListener(CompletableFuture<Boolean> future, int count) {
+        this.future = future;
+        this.count = new AtomicInteger(count);
+    }
+
+    public void notifyChangeState(int state) {
+        // Update success and fail counts
+        if (state == WORKER_MODEL_LOADED) {
+            count.decrementAndGet();
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java]: if (count.decrementAndGet() ==0) {
    future.complete(Boolean.TRUE);
}

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        }
+        if (state == WORKER_ERROR || state == WORKER_STOPPED) {
+            future.complete(Boolean.FALSE);
+        }
+        if (count.get() == 0) {
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java]: remove this block.

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 12]
```diff
@@ -110,9 +113,23 @@ public void run() {
                     throw new WorkerInitializationException(
                             "Backend worker did not respond in given time");
                 }
+                switch (req.getCommand()) {
+                    case PREDICT:
+                        model.resetFailedInfReqs();
+                        break;
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: unnecessary empty line.

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 13]
```diff
@@ -110,9 +113,23 @@ public void run() {
...
-                if (req.getCommand().equals(WorkerCommands.PREDICT)) {
-                    model.resetFailedInfReqs();
+                    case LOAD:
+                        if ("200".equals(reply.getCode())) {
+                            listener.notifyChangeState(WorkerStateListener.WORKER_MODEL_LOADED);
+                        } else {
+                            listener.notifyChangeState(WorkerStateListener.WORKER_ERROR);
+                        }
+                        break;
+
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: unnecessary empty line.

File: frontend/server/src/test/java/com/amazonaws/ml/mms/ModelServerTest.java [hunk 14]
```diff
@@ -212,6 +213,21 @@ private void testScaleModel(Channel channel) throws InterruptedException {
...
     }
 
+    private void testSyncScaleModel(Channel channel) throws InterruptedException {
+        result = null;
+        latch = new CountDownLatch(1);
+        HttpRequest req =
+                new DefaultFullHttpRequest(
+                        HttpVersion.HTTP_1_1,
+                        HttpMethod.PUT,
+                        "/models/noop_v0.1?mode=sync&min_worker=6");
```
Comment (frankfliu) [inline frontend/server/src/test/java/com/amazonaws/ml/mms/ModelServerTest.java]: Do you really need to test 6 workers?

File: frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java [hunk 15]
```diff
@@ -345,22 +346,46 @@ private void handleScaleModel(
             ChannelHandlerContext ctx, QueryStringDecoder decoder, String modelName) {
         int minWorkers = NettyUtils.getIntParameter(decoder, "min_worker", 1);
         int maxWorkers = NettyUtils.getIntParameter(decoder, "max_worker", 1);
+        String mode = NettyUtils.getParameter(decoder, "mode", null);
```
Comment (vdantu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java]: Why not have a boolean with "sync"? Can we have anything other than "sync" and "async"?

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 16]
```diff
@@ -110,9 +113,23 @@ public void run() {
...
+                    case UNLOAD:
```
Comment (vdantu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: We don't use Unload and Stats yet. Why add them here?

Comment (vrakesh) [reply frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: It is a suggestion from the ide to have them anyway

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/ModelManager.java [hunk 17]
```diff
@@ -104,17 +105,15 @@ public boolean unregisterModel(String modelName) throws WorkerInitializationExce
...
     }
 
-    public boolean updateModel(String modelName, int minWorkers, int maxWorkers)
+    public CompletableFuture<Boolean> updateModel(String modelName, int minWorkers, int maxWorkers)
             throws WorkerInitializationException {
         Model model = models.get(modelName);
         if (model == null) {
-            logger.warn("Model not found: " + modelName);
-            return false;
+            throw new WorkerInitializationException("Model not found" + modelName);
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/ModelManager.java]: missing colon and space "Model not found: ", better add error code as well.

File: frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java (hunk 15 above)
Comment (vrakesh) [reply frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java]: True, but keeping mode in a string feels more descriptive, We have two parameter methods, for key value pairs int and String, either way we do not get a boolean directly, even if the customer can pass something like sync=1 , does it prevent them from having a value more than 1?

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/ModelManager.java [hunk 18]
```diff
@@ -104,17 +105,15 @@ public boolean unregisterModel(String modelName) throws WorkerInitializationExce
...
     }
 
-    public boolean updateModel(String modelName, int minWorkers, int maxWorkers)
+    public CompletableFuture<Boolean> updateModel(String modelName, int minWorkers, int maxWorkers)
             throws WorkerInitializationException {
         Model model = models.get(modelName);
         if (model == null) {
-            logger.warn("Model not found: " + modelName);
-            return false;
+            throw new WorkerInitializationException(ErrorCodes.WORKER_INSTANTIATION_ERROR, "Model not found:" + modelName);
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/ModelManager.java]: Can you add space after "found:"

File: frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java [hunk 19]
```diff
@@ -345,22 +346,45 @@ private void handleScaleModel(
...
+        boolean async = !"sync".equalsIgnoreCase(mode);
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java]: Add this into OptionUtils

File: frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java [hunk 20]
```diff
@@ -345,22 +346,45 @@ private void handleScaleModel(
             ChannelHandlerContext ctx, QueryStringDecoder decoder, String modelName) {
         int minWorkers = NettyUtils.getIntParameter(decoder, "min_worker", 1);
         int maxWorkers = NettyUtils.getIntParameter(decoder, "max_worker", 1);
+        String synchronize = NettyUtils.getParameter(decoder, "synchronize", null);
+        boolean async = !"true".equalsIgnoreCase(synchronize);
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/http/HttpRequestHandler.java]: Better use: async = !Boolean.parseBoolean(synchronize)

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/ModelManager.java [hunk 21]
```diff
@@ -104,17 +105,16 @@ public boolean unregisterModel(String modelName) throws WorkerInitializationExce
...
 
-    public boolean updateModel(String modelName, int minWorkers, int maxWorkers)
+    public CompletableFuture<Boolean> updateModel(String modelName, int minWorkers, int maxWorkers)
             throws WorkerInitializationException {
         Model model = models.get(modelName);
         if (model == null) {
-            logger.warn("Model not found: " + modelName);
-            return false;
+            throw new WorkerInitializationException(
+                    ErrorCodes.WORKER_INSTANTIATION_ERROR, "Model not found:" + modelName);
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/ModelManager.java]: missing space

File: frontend/server/src/test/resources/open_api.txt [hunk 22]
```diff
@@ -594,6 +594,15 @@
               "type": "integer",
               "default": "-1"
             }
+          },
+          {
+            "in": "query",
+            "name": "synchronized",
```
Comment (frankfliu) [inline frontend/server/src/test/resources/open_api.txt]: synchronize

File: frontend/server/src/main/java/com/amazonaws/ml/mms/openapi/OpenApiUtils.java [hunk 23]
```diff
@@ -449,6 +449,12 @@ private static Operation getScaleOperation() {
...
                                 + " a worker to complete all pending requests. Use 0 to terminate backend"
                                 + " worker process immediately. Use -1 for wait infinitely."));
+        operation.addParameter(
+                new QueryParameter(
+                        "synchronize",
+                        "String",
+                        null,
+                        "decides whether the call to scale is synchronized or asynchronized."));
 
         operation.addResponse(new Response("202", "Accepted."));
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/openapi/OpenApiUtils.java]: Add 200 response for synchronized API

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+    private CompletableFuture<Boolean> future;
+    private AtomicInteger count;
+
+    public WorkerStateListener(CompletableFuture<Boolean> future, int count) {
+        this.future = future;
+        this.count = new AtomicInteger(count);
+    }
+
+    public void notifyChangeState(int state) {
+        // Update success and fail counts
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerStateListener.java]: Add a debug log here to troubleshoot hang issue

**Pull request merged**